### DISCUSSION TITLE
Using CopyExternalImageToTexture() to copy from webgl canvas might get blank result

### DISCUSSION THREAD

**Discussion by shaoboyan** [gpuweb/gpuweb#4356] [Ideas/Feedback]
WebGL spec says:
_"This default behavior can be changed by setting the preserveDrawingBuffer attribute of the WebGLContextAttributes object. If this flag is true, the contents of the drawing buffer shall be preserved until the author either clears or overwrites them. If this flag is false, attempting to perform operations using this context as a source image after the rendering function has returned can lead to undefined behavior. This includes readPixels or toDataURL calls, using this context as the source image of another context's texImage2D or drawImage call, or creating an ImageBitmap [HTML] from this context's canvas. "_

It means when we call `CopyExternalImageToTexture()` and use WebGL canvas as source, it __IS POSSIBLE__ that we get a webgpu texture with random contents.

Here is a example:
```
drawWebGLCanvas();
const adapter = await navigator.gpu.requestAdapter();
const device = await adapter.requestDevice();
const texture = device.createTexture({
        size: [200, 200],
        dimension: '2d',
        format: 'rgba8unorm',
        sampleCount: 1,
        usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
});
// copy from webgl canvas
device.queue.copyExternalImageToTexture({source: glCanvas}, {texture: texture}, {
        width: 200,
        height: 200
});
await device.queue.onSubmittedWorkDone();
readGPUTexture(device, texture, 1024 * texHeight);
```
We got total blank webgpu texture after running it.  

There are two possible solutions:
- Using `preserveDrawingBuffer:true` when creating webgl context. WebGL spec also confirms this.
- Copy webgl canvas content __BEFORE PRESENTING__. E.g. In Chrome, we could fix above example by moving `drawWebGLCanvas()` to the line after `const device = await adapter.requestDevice()`. This ensures copy and webgl rendering happens in the same event loop and that is "before presenting" for chrome.

So maybe we could add NOTE in webgpu spec to explain what's supposed to happen. preserveDrawingBuffer=true is needed on the WebGL canvas to make it readable after it's presented.

But I still want to know whether copy webgl canvas content __BEFORE PRESENTING__ is a spec defined thing or a browser implementation details. If it is a spec defined behavior, maybe we could document it too. If it is a browser implementation details, I hope we still could have a place to tell developers this is a possible solution.

**Comment by jimblandy** [comment]
See discussion: https://github.com/gpuweb/gpuweb/issues/4255#issuecomment-1789864754
